baybe/surrogates/base.py

```diff
@@ -56,112 +97,250 @@ class Surrogate(ABC, SerialMixin):
+            **scale-transformed** posterior distributions at the given candidate points,
+            where the posterior is described on the scale dictated by the output scaler
+            obtained via :meth:`baybe.surrogates.base.Surrogate._make_output_scaler`.
         """
 
-    def fit(self, searchspace: SearchSpace, train_x: Tensor, train_y: Tensor) -> None:
+    def fit(
+        self,
+        searchspace: SearchSpace,
+        objective: Objective,
+        measurements: pd.DataFrame,
+    ) -> None:
         """Train the surrogate model on the provided data.
 
         Args:
             searchspace: The search space in which experiments are conducted.
-            train_x: The training data points.
-            train_y: The training data labels.
+            objective: The objective to be optimized.
+            measurements: The training data in experimental representation.
 
         Raises:
             ValueError: If the search space contains task parameters but the selected
                 surrogate model type does not support transfer learning.
             NotImplementedError: When using a continuous search space and a non-GP
                 model.
         """
+        # TODO: consider adding a validation step for `measurements`
+
         # Check if transfer learning capabilities are needed
         if (searchspace.n_tasks > 1) and (not self.supports_transfer_learning):
             raise ValueError(
                 f"The search space contains task parameters but the selected "
                 f"surrogate model type ({self.__class__.__name__}) does not "
                 f"support transfer learning."
             )
-        # TODO: Adjust scale_model decorator to support other model types as well.
         if (not searchspace.continuous.is_empty) and (
             "GaussianProcess" not in self.__class__.__name__
         ):
             raise NotImplementedError(
                 "Continuous search spaces are currently only supported by GPs."
             )
 
-        # Validate and prepare the training data
-        train_x = _prepare_inputs(train_x)
-        train_y = _prepare_targets(train_y)
+        # Remember on which search space the model is trained
+        self._searchspace = searchspace
+
+        # Create context-specific transformations
+        self._input_scaler = self._make_input_scaler(searchspace)
+        self._output_scaler = self._make_output_scaler(objective, measurements)
 
-        self._fit(searchspace, train_x, train_y)
+        # Transform and fit
+        train_x_comp_rep, train_y_comp_rep = to_tensor(
+            searchspace.transform(measurements, allow_extra=True),
+            objective.transform(measurements),
+        )
+        train_x = self._input_scaler.transform(train_x_comp_rep)
+        train_y = (
+            train_y_comp_rep
+            if self._output_scaler is _IDENTITY_TRANSFORM
+            else self._output_scaler(train_y_comp_rep)[0]
```

Comment:
Right now, this stems from the fact that:
* we have several input dimensions
* but only one output dimension

When there is no output scaler defined, you can skip the transformation altogether (and have to, because there exists no scaler object in that case).

For the input scaler, you could only skip the transformation if ALL columns skip transformation. In fact, we *could* change the logic a little bit in the sense that we also return some `_IDENTITY_TRANSFORM` sentinel when no columns are transformed – right now, we do an actual transformation of columns with an all-column identity transformer. If you want me to change that, let me know.

Alternatively, we can also change this in October, when we enable multi-target. Because then the target part needs to change anyway and the scaling logic needs to be checked again. What do you prefer? Or anything else?

Reply:
I didn't mean that, I meant that in one case you call `self._output_scaler(train_y_comp_rep)[0]` and in the other `self._input_scaler.transform(train_x_comp_rep)`

Reply:
Ah I see. That's currently a consequence of the (rather arbitrary) design. For the output scaler, there is not real choice since we're using directly the botorch transforms. Of course, we could wrap them to impose a different interface, but no real reason to do so ... For the input scaler, it's simply because I've implemented the custom column scaler in the scikit-learn style, and not in the torch style. Probably, we should rather use the later for consistency reasons, but how about we change this as soon as we introduce multi-target? Because then the target scaler needs wrapping as well

Reply:
ok but especially if you know one of them will have to change anyway rather soon to the other version then it might make sense to already fix the inconsistency now

Reply:
Looking at the time left before vacation, I'd actually prefer changing it later when I touch the code anyway. Functionally, everything would stay as is but would require extending the review process again, etc... Let us close this PR now and rather get the remaining things merged and released before I leave 👍🏼